Convert filterpipeline deserialize to factory

bdeng-xt opened the pull request:
This PR converted deserialize function of FilterPipeline class to static function.

---
TYPE:  IMPROVEMENT 
DESC: Convert FilterPipeline deserialize function to static factory function

eric-hughes-tiledb commented on tiledb/sm/filter/filter_pipeline.h:
```diff
@@ -92,9 +101,10 @@ class FilterPipeline {
    */
-  Status deserialize(ConstBuffer* buff);
+  static std::tuple<Status, optional<std::shared_ptr<FilterPipeline>>>
+  deserialize(ConstBuffer* buff);
```
I don't think the `shared_ptr` is extraneous here. Every place that's calling `deserialize` is immediately deferencing the pointer (with `operator*()`). The intermediate allocation doesn't seem to be doing anything.

bdeng-xt replied:
changed to return std::tuple<Status, optional<FilterPipeline>>

eric-hughes-tiledb commented on tiledb/sm/array_schema/array_schema.cc:
```diff
@@ -570,14 +570,30 @@ Status ArraySchema::deserialize(ConstBuffer* buff) {
+  if (!st_coords_filters.ok()) {
+    return Status_ArraySchemaError("Cannot deserialize coords filters");
+  }
+  coords_filters_ = FilterPipeline(*coords_filters.value());
```
You shouldn't need to explicitly call a constructor here. The assignment operator should work fine.

eric-hughes-tiledb commented on tiledb/sm/array_schema/array_schema.cc:
```diff
@@ -570,14 +570,30 @@ Status ArraySchema::deserialize(ConstBuffer* buff) {
+      return Status_ArraySchemaError(
+          "Cannot deserialize cell validity filters");
+    }
+    cell_validity_filters_ = FilterPipeline(*cell_validity_filters.value());
```
same issue

eric-hughes-tiledb commented on tiledb/sm/filter/filter_pipeline.h:
```diff
@@ -255,7 +265,7 @@ class FilterPipeline {
 
   /** The ordered list of filters comprising the pipeline. */
-  std::vector<tdb_unique_ptr<Filter>> filters_;
+  std::vector<std::shared_ptr<Filter>> filters_;
```
Is there a particular reason to switch from unique to shared here?

bdeng-xt replied:
That is because FilterCreate::deserialize returns shared_ptr. and shared_ptr can not convert to uniqu_ptr(https://stackoverflow.com/questions/37884728/does-c11-unique-ptr-and-shared-ptr-able-to-convert-to-each-others-type)

bdeng-xt replied:
That is because FilterCreate::deserialize returns shared_ptr. and shared_ptr can not convert to uniqu_ptr(https://stackoverflow.com/questions/37884728/does-c11-unique-ptr-and-shared-ptr-able-to-convert-to-each-others-type)

eric-hughes-tiledb reviewed (changes requested):
Looks fine but for some removal of `std::` qualifications: `tuple`, `optional`, `shared_ptr`. I've not marked them individually; there are lots, and it would be better to use a search tool to find them all. This is a good time to remove all the occurrences of them in the files touched by this PR. Many are in new code; some might be there already and would add to this PR.

bdeng-xt commented:
> Looks fine but for some removal of `std::` qualifications: `tuple`, `optional`, `shared_ptr`. I've not marked them individually; there are lots, and it would be better to use a search tool to find them all. This is a good time to remove all the occurrences of them in the files touched by this PR. Many are in new code; some might be there already and would add to this PR.

Removed `std::` and rebased to dev

eric-hughes-tiledb reviewed (approved):
LGTM
